doc: add vm-history in hardforks (#499)
nervosnetwork/docs.nervos.org: website/docs/history-and-hard-forks/history-vm-version.mdx

---
id: history-vm-version
title: "History VM Version"
---

> Usually, a hard fork will bring a new CKB-VM version in CKB network history for better security, performance, bug fixings and new RISC-V extensions. This article lists all the version of CKB-VM from previous hard fork history.
> 
> One thing to be noted is that, even the hard fork brings upgrade to the network, it is important to prevent the hard fork from breaking the old Script codes. CKB makes best efforts to ensure users have the opt-in option to specify the VM version for their Script. For the specific mechanism that determines how the CKB node chooses the CKB-VM version for a transaction Script group, please refer to [vm-selection](/docs/script/vm-selection).

Reviewer: ```suggestion
The CKB network has introduced various CKB-VM versions over time to enhance security, performance, resolve bugs, and support new RISC-V extensions. This article lists all versions of the CKB-VM from previous hard fork events.

It is important to note that, even though a hard fork upgrades the network, it is crucial to prevent the hard fork from breaking the old Script codes. CKB makes best effort to ensure users have the opt-in option to specify the VM version for their Script. For the specific mechanism that determines how the CKB node chooses the CKB-VM version for a transaction's Script group, please refer to [vm-selection](/docs/script/vm-selection).
```

- [Load Cell Data](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0009-vm-syscalls/0009-vm-syscalls.md#load-cell-data)
- [Load Cell Data As Code](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0009-vm-syscalls/0009-vm-syscalls.md#load-cell-data-as_code)
- [Load Input](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0009-vm-syscalls/0009-vm-syscalls.md#load-input)
- [Load Input By Field](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0009-vm-syscalls/0009-vm-syscalls.md#load-input-by-field)
- [Load Header](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0009-vm-syscalls/0009-vm-syscalls.md#load-header)
- [Load Header By Field](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0009-vm-syscalls/0009-vm-syscalls.md#load-header-by-field)
- [Load Witness](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0009-vm-syscalls/0009-vm-syscalls.md#load-witness)
- [Debug](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0009-vm-syscalls/0009-vm-syscalls.md#debug)

### Cost Model

Full description: [RFC](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0014-vm-cycle-limits/0014-vm-cycle-limits.md)

> Specify the cycles needed by each CKB VM instructions or syscalls.
> 
> 1. Cycles for RISC-V instructions are determined based on real hardware that implement RISC-V ISA.
> 2. Cycles for syscalls are measured based on real runtime performance metrics obtained while benchmarking current CKB implementation.
> 
> **Initial Loading Cycles**
> 
> For each byte loaded into CKB VM in the initial ELF loading phase, 0.25 cycles will be charged. This is to encourage dapp developers to ship smaller smart contracts as well as preventing DDoS attacks using large binaries. Notice fractions will be rounded up here, so 30.25 cycles will become 31 cycles.

Reviewer: ```suggestion
Specify the cycles needed by each CKB-VM instructions or syscalls.

1. Cycles for RISC-V instructions are determined based on real hardware that implement RISC-V ISA.
2. Cycles for syscalls are measured based on real runtime performance metrics obtained while benchmarking current CKB implementation.

**Initial Loading Cycles**

For each byte loaded into CKB-VM in the initial ELF loading phase, 0.25 cycles will be charged. This is to encourage dApp developers to ship smaller Scripts as well as preventing DDoS attacks using large binaries. Notice fractions will be rounded up here, so 30.25 cycles will become 31 cycles.
```

| LB          | 3                    |
| LWU         | 3                    |
| LHU         | 3                    |
| LBU         | 3                    |
| SW          | 3                    |
| SH          | 3                    |
| SB          | 3                    |
| LWSP        | 3                    |
| SWSP        | 3                    |
| MUL         | 5                    |
| MULW        | 5                    |
| MULH        | 5                    |
| MULHU       | 5                    |
| MULHSU      | 5                    |
| DIV         | 32                   |
| DIVW        | 32                   |
| DIVU        | 32                   |
| DIVUW       | 32                   |
| REM         | 32                   |
| REMW        | 32                   |
| REMU        | 32                   |
| REMUW       | 32                   |
| ECALL       | 500 (see note below) |
| EBREAK      | 500 (see note below) |

**Syscall Cycles**

As shown in the above chart, each syscall will have 500 initial cycle consumptions. This is based on real performance metrics gathered benchmarking CKB implementation, certain bookkeeping logics are required for each syscall here.

> In addition, for each byte loaded into CKB VM in the syscalls, 0.25 cycles will be charged. Notice fractions will also be rounded up here, so 30.25 cycles will become 31 cycles.

Reviewer: ```suggestion
In addition, for each byte loaded into CKB-VM in the syscalls, 0.25 cycles will be charged. Notice fractions will also be rounded up here, so 30.25 cycles will become 31 cycles.
```

## VM 1

> introduced from **1st Hard Fork – CKB Edition Mirana (2021)**

Reviewer: ```suggestion
Introduced since [1st Hard Fork – CKB Edition Mirana (2021)](/docs/history-and-hard-forks/ckb-hard-fork-history#1st-hard-fork--ckb-edition-mirana-2021)
```

| ADC [2](https://github.com/nervosnetwork/ckb-vm/issues/169) | add + sltu + add + sltu + or | 1 + 0 + 0 + 0 + 0 |
| SBB                                                         | sub + sltu + sub + sltu + or | 1 + 0 + 0 + 0 + 0 |
| WIDE_MUL                                                    | mulh + mul                   | 5 + 0             |
| WIDE_MULU                                                   | mulhu + mul                  | 5 + 0             |
| WIDE_MULSU                                                  | mulhsu + mul                 | 5 + 0             |
| WIDE_DIV                                                    | div + rem                    | 32 + 0            |
| WIDE_DIVU                                                   | divu + remu                  | 32 + 0            |
| FAR_JUMP_REL                                                | auipc + jalr                 | 0 + 3             |
| FAR_JUMP_ABS                                                | lui + jalr                   | 0 + 3             |
| LD_SIGN_EXTENDED_32_CONSTANT                                | lui + addiw                  | 1 + 0             |

## VM 2

> **introduced from 2nd Hard Fork – CKB Edition Meepo (2024)**

Reviewer: ```suggestion
Introduced since [2nd Hard Fork – CKB Edition Meepo (2024)](/docs/history-and-hard-forks/ckb-hard-fork-history#2nd-hard-fork--ckb-edition-meepo-2024)
```
